src/Symfony/Component/HttpClient/Response/ApiResponse.php

```diff
@@ -0,0 +1,169 @@
+    private $timeout;
+
+    /**
+     * @internal
+     */
+    public function __construct(ResponseInterface $response)
+    {
+        $this->response = $response;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getStatusCode(): int
+    {
+        $this->clearError();
+
+        return $this->response->getStatusCode();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getHeaders(bool $throw = true): array
+    {
+        $this->clearError();
+
+        return $this->response->getHeaders($throw);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getContent(bool $throw = true): string
+    {
+        $this->clearError();
+
+        return $this->response->getContent($throw);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getInfo(string $type = null)
+    {
+        return $this->response->getInfo($type);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toArray(bool $throw = true): array
+    {
+        $this->clearError();
+        $contentType = $this->response->getHeaders($throw)['content-type'][0] ?? 'application/json';
+
+        if (null !== $this->data) {
+            return $this->data;
+        }
+
+        if ('' === $data = $this->response->getContent(false)) {
+            throw new TransportException('API returned empty response body.');
+        }
+
+        if (!preg_match('/\bjson\b/i', $contentType)) {
+            throw new JsonException(sprintf('API returned content-type "%s" while a JSON-compatible one was expected.', $contentType));
+        }
+
+        try {
+            $data = json_decode($data, true, 512, JSON_BIGINT_AS_STRING | (\PHP_VERSION_ID >= 70300 ? JSON_THROW_ON_ERROR : 0));
+        } catch (\JsonException $e) {
+            throw new JsonException($e->getMessage(), $e->getCode(), $e);
+        }
+
+        if (\PHP_VERSION_ID < 70300 && JSON_ERROR_NONE !== json_last_error()) {
+            throw new JsonException(json_last_error_msg(), json_last_error());
+        }
+
+        if (!\is_array($data)) {
+            throw new JsonException(sprintf('API returned %s while an array was expected.', \gettype($data)));
+        }
+
+        return $this->data = $data;
+    }
+
+    public function __destruct()
+    {
+        if ($this->timeout) {
+            throw new TransportException('API response reached the inactivity timeout.');
+        }
+
+        $this->clearError();
+    }
+
+    private function clearError()
+    {
+        $e = $this->error;
+        $this->timeout = $this->error = null;
+
+        if ($e) {
+            throw new TransportException($e->getMessage(), 0, $e);
+        }
+    }
+
+    /**
+     * @internal
+     */
+    public static function complete(HttpClientInterface $client, iterable $apiResponses, ?float $timeout): \Generator
+    {
+        $responses = new \SplObjectStorage();
+
+        foreach ($apiResponses as $r) {
+            $responses[$r->response] = $r;
```

Comment:
Sure, documented, and now check with proper exception. Only ApiResponse are accepted because that's the design of the component: the responses are not standalone object, they're bound by "friendship" with the client that emitted them. That's how clients know how to stream of complete them.

Reply:
That seems like break of contract of `ApiClientInterface`. `ApiClientInterface::complete` says you can use `ApiResponseInterface`, but `ApiClient::complete` rejects everything that is not `ApiResponse`.

Reply:
The contract is on `ApiClientInterface::complete` not `ApiResponseInterface::complete`

Reply:
It's very alike PSR 6 cache items: they are not interoperable, like responses are not interoperable from one client to another.

Here is how PSR 6 tells it:

>  Calling Libraries SHOULD NOT assume that an Item created by one Implementing Library is compatible with a Pool from another Implementing Library.

Same for responses.

Reply:
I wonder why not design fitting contract? We are in design phase. Contract should declare everything implementations need. If even main implementation works around the contract, that's not a good signal.

Reply:
But a contract is not only about type hints. If the contracts says that responses cannot be passed from one client implementation to another, that's a the contracts, and a perfectly valid one. Exactly as PSR 6.

Reply:
Rereading, I may not have answered your question @ostrolucky.

> Contract should declare everything implementations need

That's actually what happens here. Tell me if I'm wrong, but I understand from your comment that you think we should look for completely standalone response classes: we should design each interface as a standalone contract (eg responses) that could we reused independently with any other interface in the domain (eg clients). What I'm trying to explain above is that a class is not the only possible unit to define the boundaries of reusable code units. Here, like in PSR6, the client is the only "legal" factory for response objects. For this reason, we do not need to define any contracts to make them talk to each other: that communication is purely internal details.

Another way to look at the problem is that one: do we need to work on a set of interfaces that would allow any response objects to work with any clients? Let's be more concrete: do we need to define interfaces that would allow a response created by a client based on **curl** to be streamed by a client based on **fopen**?

Maybe it's possible, by adding more abstractions in between. But which problem would this solve in practice? Nobody will ever need to do that.

So, in conclusion, the client+response form a tuple that are bound together by some internal detail the outside world should not care about. That's actually the design decision that makes the contracts work so seamlessly and so consistently.

Reply:
> contracts says that responses cannot be passed from one client implementation to another

1. Can you point me to a part of the contract in http-client which says this? I don't see it.
1. Even if contract says that, how does this non-interoperability benefit user?

> That's actually the design decision that makes the contracts work so seamlessly and so consistently.

My impression is it makes whole thing work less seamlessly and less consistently. Even you had to workaround your own contract with SplObjectStorage, because http client implementation rejects any request it didn't itself create. Anybody creating their own implementation will have to jump through same hoops. And it has code smells like methods in response accepting other responses and task scheduling bult in, instead of in client or separate class. This is less likely to happen with contracts defining boundaries properly, like having ResponseInterface::isFinished

> Exactly as PSR 6.

PSR-6 doesn't forbid anybody to make pools work with items from other pools. It's unfortunate that symfony/cache chose this option. symfony/http-client goes even more overboard and doesn't even reuse same Response (Item) class across Clients (ItemPools). Imagine how much more work you would have if every cache adapter came with its own item class. This is happening here, just nobody realizes because there isn't many implementations (yet).


> that communication is purely internal details

This is leaky internal detail and will surface when trying to implement own Response class. And contract doesn't warn even in text form. Not even clients do that. Every client pretends it accepts interface responses, while deeper layer that it uses rejects everything else than concrete, final class.

> do we need to define interfaces that would allow a response created by a client based on curl to be streamed by a client based on fopen?

Http-client goes way more far than that. It's not that consumer is not able to use curl response in fopen client, but they are not able to use ANY other response. They cannot even use wrapped curl response in curl client, or curl response subclass. If consumer wants to create new Response, they must create new client as well. That includes cases when consumer would wish to add just one utility method. Combined with the fact everything has to be decorator because nothing is inheritable, this looks too verbose. And when user begins to have a need to have own response classes, they can no longer safely use interface, because anything else than their own implementation might blow up when passed to `ApiClientInterface::complete` or `HttpClientInterface::stream`, better to typehint to their own client instead across whole application to avoid can of worms. Even then, user must be careful each time passing these to external libraries. Maybe I am overreacting because this applies only to those mentioned bulk methods (and their use will be rare), but this will be PITA when somebody will need to use those in this context.

> But which problem would this solve in practice? Nobody will ever need to do that.

I might wish to create: XmlResponse, CsvResponse, StripeResponse, StringeableResponse, MockResponse, UserCreateResponse, TransactionAwareResponse, TraceableResponse etc. I might wish to use unique response class for every API endpoint. Use case this solves is preserve ability to safely pass these to `ApiClientInterface::complete` or `HttpClientInterface::stream` without unwinding & rewinding these response wrappers (which cannot even happen without instanceof checks, because they can be passed to different layers)

Reply:
Relates to https://github.com/symfony/symfony/pull/30413#issuecomment-468828899